### PR TITLE
ENH: Add annotations for `np.testing`

### PR DESCRIPTION
**Pull request by BvB93**
This PR adds annotations for all public(-ish?) objects in `np.testing`.

### CODE REVIEW

File: numpy/testing/_private/utils.pyi [hunk 1]
```diff
@@ -0,0 +1,396 @@
...
+        __exc_tb: None | types.TracebackType = ...,
+    ) -> None: ...
+    def __call__(self, func: _FT) -> _FT: ...
+
+verbose: int
+IS_PYPY: Final[bool]
+HAS_REFCOUNT: Final[bool]
+HAS_LAPACK64: Final[bool]
+
+def assert_(val: object, msg: str | Callable[[], str] = ...) -> None: ...
```
Comment (BvB93) [inline numpy/testing/_private/utils.pyi:141]: `msg` might potentially be a bit on the strict side here, as the `AssertionError` constructor can in principle take any object.

File: numpy/testing/_private/utils.pyi [hunk 2]
```diff
@@ -0,0 +1,396 @@
...
+    cls: Type[Any],
+    decorator: Callable[[Callable[..., Any]], Any],
+    testmatch: None | str | bytes | Pattern[Any] = ...,
+) -> None: ...
+
+def measure(
+    code_str: str | bytes | ast.mod | ast.AST,
+    times: int = ...,
+    label: None | str = ...,
+) -> float: ...
```
Comment (BvB93) [inline numpy/testing/_private/utils.pyi:302]: There are a bunch of functions within this module that are thin wrappers around standard library functions.
In such cases I, more or less, copied their respective annotations from [typeshed](https://github.com/python/typeshed).

File: numpy/testing/utils.py [hunk 3]
```diff
@@ -24,5 +25,5 @@
         'assert_allclose', 'IgnoreException', 'clear_and_catch_warnings',
         'SkipTest', 'KnownFailureException', 'temppath', 'tempdir', 'IS_PYPY',
         'HAS_REFCOUNT', 'suppress_warnings', 'assert_array_compare',
-        '_assert_valid_refcount', '_gen_alignment_data', 'assert_no_gc_cycles'
```
Comment (BvB93) [inline numpy/testing/utils.py:27]: There seem to be a number of objects within `__all__` with a rather questionable public status, these two being the most obvious ones. Deprecation might be overkill, but doing a review and removing some of the offending objects from `__all__` would be a good start.

File: numpy/testing/_private/utils.py [hunk 4]
```diff
@@ -35,8 +35,7 @@
         'assert_allclose', 'IgnoreException', 'clear_and_catch_warnings',
         'SkipTest', 'KnownFailureException', 'temppath', 'tempdir', 'IS_PYPY',
         'HAS_REFCOUNT', 'suppress_warnings', 'assert_array_compare',
-        '_assert_valid_refcount', '_gen_alignment_data', 'assert_no_gc_cycles',
-        'break_cycles', 'HAS_LAPACK64'
+        'assert_no_gc_cycles', 'break_cycles', 'HAS_LAPACK64'
```
Comment (charris) [inline numpy/testing/_private/utils.py:38]: The private functions are actually used and imported directly from `numpy.testing`, which in turn imports them with `*`.

File: numpy/testing/utils.py (hunk 3 above)
Comment (charris) [reply numpy/testing/utils.py:27]: Note that we will remove `numpy/testing/utils.y` at some point. It there for backwards compatibility and to warn downstream to change the import.

File: numpy/testing/_private/utils.py (hunk 4 above)
Comment (charris) [reply numpy/testing/_private/utils.py:38]: IIRC, they are due to @seberg . I've wondered about their private status.

Comment (BvB93) [reply numpy/testing/_private/utils.py:38]: Note that respective private functions are still imported explicitly to `np.testing.__init__` and `np.testing.utils`, 
so the content of these namespaces hasn't changed.
They're just no longer part of `__all__`, as putting referencing private objects within the `__all__` of a public module is not exactly great practice.

**Comment by charris** [comment]
Thanks Bas.

**Pull request merged**